src/main/scala/com/monsanto/arch/cloudformation/model/resource/EC2.scala

```diff
@@ -109,14 +109,15 @@ object `AWS::EC2::CustomerGateway` extends DefaultJsonProtocol {
 }
 
 @implicitNotFound("A Route can only have exactly ONE of GatewayId, InstanceId, NetworkInterfaceId or VpcPeeringConnectionId set")
-class ValidRouteCombo[G, I, P] private ()
-object ValidRouteCombo{
-  implicit object valid1T extends ValidRouteCombo[Some[Token[ResourceRef[`AWS::EC2::InternetGateway`]]], None.type, None.type]
-  implicit object valid1 extends ValidRouteCombo[Some[ResourceRef[`AWS::EC2::InternetGateway`]], None.type, None.type]
-  implicit object valid2T extends ValidRouteCombo[None.type , Some[Token[ResourceRef[`AWS::EC2::Instance`]]], None.type]
-  implicit object valid2 extends ValidRouteCombo[None.type , Some[ResourceRef[`AWS::EC2::Instance`]], None.type]
-  implicit object valid3T extends ValidRouteCombo[None.type , None.type, Some[Token[ResourceRef[`AWS::EC2::VPCPeeringConnection`]]]]
-  implicit object valid3 extends ValidRouteCombo[None.type , None.type, Some[ResourceRef[`AWS::EC2::VPCPeeringConnection`]]]
+sealed trait ValidRouteComboOption
+case class InternetGatewayRoute(v:Token[ResourceRef[`AWS::EC2::InternetGateway`]]) extends ValidRouteComboOption
+case class EC2InstanceRoute(v:Token[ResourceRef[`AWS::EC2::Instance`]]) extends ValidRouteComboOption
+case class VPCPeeringRoute(v:Token[ResourceRef[`AWS::EC2::VPCPeeringConnection`]]) extends ValidRouteComboOption
+
+object ValidRouteComboOption {
+  implicit def toInternetGateway[T](v: T)(implicit t:T => Token[ResourceRef[`AWS::EC2::InternetGateway`]]) = InternetGatewayRoute(v)
```

Comment:
Implicit view resolution is only recursive if the implicit view itself needs another implicit. So I do it for chaining. In this case, we need chaining: We turn a Resource into a Token[ResourceRef] with an implicit, and we want to combine them both for free.

http://docs.scala-lang.org/tutorials/FAQ/chaining-implicits.html